core/shared/src/main/scala/zio/ZIOApp.scala

```diff
@@ -101,7 +101,7 @@ trait ZIOApp extends ZIOAppPlatformSpecific with ZIOAppVersionSpecific { self =>
           () => runtime.unsafe.run(Fiber.dumpAll)(trace, Unsafe.unsafe).getOrThrowFiberFailure()(Unsafe.unsafe)
 
         if (System.os.isWindows) {
-          Platform.addSignalHandler("INT", dumpFibers)(Unsafe.unsafe)
+          Platform.addSignalHandler("SIGINT", dumpFibers)(Unsafe.unsafe)
```

Comment:
using `TERM` also works

Reply:
This is weird. I think that would actually make the handler just not work. 

Look at jdk.internal.misc

```
    public Signal(String name) {
        Objects.requireNonNull(name, "name");
        // Signal names are the short names;
        // the "SIG" prefix is not used for compatibility with previous JDKs
        if (name.startsWith("SIG")) {
            throw new IllegalArgumentException("Unknown signal: " + name);
        }
        this.name = name;
        number = findSignal0(name);
        if (number < 0) {
            throw new IllegalArgumentException("Unknown signal: " + name);
        }
    }
```

The javadoc at the top of the file specifically shows calling it with the "INT" string, not "SIGINT"

```
This class provides ANSI/ ISO C signal support. A Java program can register signal handlers for the current process. There are two restrictions:
Java code cannot register a handler for signals that are already used by the Java VM implementation. The Signal. handle function raises an IllegalArgumentException if such an attempt is made.
When Signal. handle is called, the VM internally registers a special C signal handler. There is no way to force the Java signal handler to run synchronously before the C signal handler returns. Instead, when the VM receives a signal, the special C signal handler creates a new thread (at priority Thread. MAX_PRIORITY) to run the registered Java signal handler. The C signal handler immediately returns. Note that because the Java signal handler runs in a newly created thread, it may not actually be executed until some time after the C signal handler returns.
Signal objects are created based on their names. For example:
new Signal("INT");
constructs a signal object corresponding to SIGINT, which is typically produced when the user presses Ctrl-C at the command line. The Signal constructor throws IllegalArgumentException when it is passed an unknown signal.
This is an example of how Java code handles SIGINT:
Signal.Handler handler = new Signal. Handler () {     public void handle(Signal sig) {       ... // handle SIGINT     } }; Signal. handle(new Signal("INT"), handler);
Since:
9
```

Reply:
I don't understand how, but it works on my PC - do you think using scala reflection has something to do with it?

Reply:
Makes me wonder whether the part that makes it work is that we're adding at least 1 signal handler. Does this work with other signal handlers or just `SIGINT`?

Also, you could probably add some debugging to check whether any of the error handlers are catching errors when handling the signal in Windows

Reply:
I think that this fixed the problem by introducing a bug =)

Before merging you should see if it does the same thing by not registering the SIGINT handler at all.

It looks to me like registering a signal handler for SIGINT, makes it so the JVM then does not run the shutdown hooks, if the SIGINT handler does not shut down the JVM. This lets the application decide if it wants to keep running or not.

Look at these examples in plain Scala. The first correctly runs the shutdown hook, and exits with a 130.

```
object Hook extends App:

  val runningThread = new Thread:
    override def run(): Unit =
      try
        while true do
          println("Running")
          sleep(1000)
      catch
        case _: InterruptedException =>
          println("Sleep Interrupted")
          Thread.currentThread().interrupt()

  val shutdownHookThread = new Thread:
    override def run(): Unit =
      println("Shutting down")
      runningThread.interrupt()
      runningThread.join()

  java.lang.Runtime.getRuntime.addShutdownHook(shutdownHookThread)

  runningThread.start()
```

This second example, registers a SIGINT handler, and when run, the handler is called, but the shutdown hook is not, and the JVM keeps running.

```
object Hook extends App:

  val handler = new SignalHandler():
    override def handle(signal: Signal): Unit =
      println("Signal " + signal)

  val runningThread = new Thread:
    override def run(): Unit =
      try
        while true do
          println("Running")
          sleep(1000)
      catch
        case _: InterruptedException =>
          println("Sleep Interrupted")
          Thread.currentThread().interrupt()

  val shutdownHookThread = new Thread:
    override def run(): Unit =
      println("Shutting down")
      runningThread.interrupt()
      runningThread.join()

  Signal.handle(new Signal("INT"), handler);

  java.lang.Runtime.getRuntime.addShutdownHook(shutdownHookThread)

  runningThread.start()
```

So I think in the pull request, trying to register using the string "SIGINT" instead of "INT" makes it so the signal handler registration fails, but this fixes the issue, because the shutdown hook still runs.

Before, when the correct signal handler, registered with the "INT" string runs, then the signal handler is fired, which just dumps the fibers to the console, but the shutdown hook doesn't run.

So either we should just not register for SIGINT at all, or if we do, we need to do more than just dump the fibers, we need to actually request the runtime to shutdown.

Reply:
One final example, this time in the signal handler, we ask the system to shutdown, and we get both the signal handler running, and the shutdown hook running, and the system exits cleanly, this time with a 0 exit code, rather than the 130, because we handled the 130 signal, and decided to shutdown. We could exit with any code we want here, or choose not to exit also.

```
object Hook extends App:

  val handler = new SignalHandler():
    override def handle(signal: Signal): Unit =
      println("Signal " + signal)
      println("Shutting down via signal handler")
      java.lang.System.exit(0)

  val runningThread = new Thread:
    override def run(): Unit =
      try
        while true do
          println("Running")
          sleep(1000)
      catch
        case _: InterruptedException =>
          println("Sleep Interrupted")
          Thread.currentThread().interrupt()

  val shutdownHookThread = new Thread:
    override def run(): Unit =
      println("Shutdown Hook")
      runningThread.interrupt()
      runningThread.join()

  Signal.handle(new Signal("INT"), handler);

  java.lang.Runtime.getRuntime.addShutdownHook(shutdownHookThread)

  runningThread.start()
```

Reply:
So I guess my final conclusion here, is that for me at least, it looks like the existing code is correct, but it is not desired behavior.

I feel that SIGINT is to interrupt, and our IDE's and the Windows Service Manager seem to agree, but our code is treating it like a SIGINFO or SIGUSR1 which are not designed to interrupt, but to ask the process for it's status, or signal it for some other reason (user defined).

So I would just ask that on Windows, we don't dump the fibers and continue, but that we exit. That's the behavior if we don't register the signal handler.

So the fix would be just to do something like this in ZIOApp:

```
  protected def installSignalHandlers(runtime: Runtime[Any])(implicit trace: Trace): UIO[Any] =
    ZIO.attempt {
      if (!ZIOApp.installedSignals.getAndSet(true)) {
        val dumpFibers =
          () => runtime.unsafe.run(Fiber.dumpAll)(trace, Unsafe.unsafe).getOrThrowFiberFailure()(Unsafe.unsafe)

        if (!System.os.isWindows) {
          Platform.addSignalHandler("INFO", dumpFibers)(Unsafe.unsafe)
          Platform.addSignalHandler("USR1", dumpFibers)(Unsafe.unsafe)
        }
      }
    }.ignore
```

Reply:
@mlhales kudos to you for providing great examples which helped me have a better understanding on whats actually happening
that's what was bugging me when I was browsing how typelevel does it in cats effect, they also seem to not register the signal handler for windows
https://github.com/typelevel/cats-effect/blob/28f8f294070412fe829f41c79e864e49a03a6299/core/jvm/src/main/scala/cats/effect/IOApp.scala#L441